enhancements/sig-architecture/70-spread-policy/README.md

```diff
@@ -0,0 +1,301 @@
+
+**Example.** We have a backend service to deploy to the public cloud. We prefer aws than other vendors for some reasons. The scheduler will first try to select only aws clusters then clusters from other vendors only if aws cannot provide enough clusters.
+
+### Use Case 5: Joint Spread
+
+To express hierarchical topology structure, clusters are labeled with multiple topology keys (i.e., region, zone, etc.), we want them to work jointly to determine the spread decision.
+
+**Example.** We want four clusters from aws to put our backend service. Further, we want an even spread across regions and an affinity spread across zones. Specifically, for the region us-east-1, we want to first put the workload to us-east-1a then us-east-1b. Assuming all zones here can provide enough clusters. Then, the scheduling result should contain two clusters from us-east-1a and other two clusters from us-west-1a.
+
+### Use Case 6: Spread with Ratio (Out of Scope)
+
+Sometimes, we may need to spread the workload with a specified ratio. Let us consider a scenario where most of the access to the workload is from the east coast of the US, even spread between us-east-1 and us-west-1 would be inappropriate. Instead, we want to place more workload on us-east-1 than us-west-1. Besides, we want to control such preferences by user-specified configurations.
+
+**Example.** We want six clusters to deploy our backend service. Moreover, we want 2/3 of them to be located in us-east-1 and 1/3 to be located in us-west-1 since there are more clients from the east coast of the US.
+
+## Goal
+Generally, we have two goals for the spread policy:
+1. Cover the full functionality in the Motivation section;
+2. Implement a high-performance and scalable mechanism that works for thousands of clusters efficiently.
+
+## Proposal
+
+### API Spec
+```go
+type PlacementSpec struct {
+	// ...
+
+	// SpreadConstraints defines how placement decision should be distributed among a
+	// set of clusters.
+	SpreadConstraints []SpreadConstraintsTerm `json:"spreadConstraints,omitempty"`
+}
+
+// SpreadConstraintsTerm defines a terminology to spread placement decisions
+type SpreadConstraintsTerm struct {
+	// Type is the type of SpreadConstraintsTerm. It could be Even or Affinity.
+	// +required
+	Type string `json:"type"`
+
+	// TopologyKey is either a label key or a cluster claim name of ManagedClusters
+	// +required
+	TopologyKey string `json:"topologyKey"`
+
+	// TopologyKeyType indicates the type of TopologyKey. It could be Label or Claim.
+	// +required
+	TopologyKeyType string `json:"topologyKeyType"`
+
+	// TopologyWeights indicates the optional weights of topologies, valid when Type
+	// is Affinity.
+	TopologyWeights []TopologyWeightTerm `json:"topologyWeights,omitempty"`
```

Comment:
I think this should be handled differently. I think what you want is to put as many as possible decisions in the certain set of clusters. Mixing this with spreading is confusing.

Reply:
The initial rationale of mixing affinity and even spreading is to force users to think and write the Placement API following the TopologyKey. If we separating affinity and even spreading, a user might submit an object like:

```
apiVersion: cluster.open-cluster-management.io/v1alpha1
kind: Placement
metadata:
  name: demo-placement
spec:
  # ...
  affinity:
    clusterAffinity:
      labelSelector:
        matchExpressions:
        - key: provider
          operator: In
          values:
          - AWS
  spreadConstraints:
  - topologyKey: provider
    topologyKeyType: Label
  # ...
```
If we have both AWS and GCP clusters, it is confusing to both the scheduler and the user himself/herself, as the spread constraint and the affinity rule specify opposite semantics towards the same topologyKey.

By mixing affinity and even spreading, the SpreadConstraintsTerms in the SpreadConstraints array will be ensured to have different topologyKeys using api validation. As each SpreadConstraintsTerm can be of either Even OR Affinity type, the problem may be solved.

I'm aware that such a design seems complex and differs with K8s's API design. If my motivation is unreal, affinity and even spreading can be certainly separated. And This proposal might focus on even spreading and put affinity out of scope.

Reply:
I think we should separate them. I feel that the API validation will be similar and it is complicated when we need to treat matchExpression and even more complicated scenario (e.g. including predicate). This might be surfaced by events or condition in the placement instead of using validation.

Reply:
OK, I'll edit the proposal and take your suggestions.